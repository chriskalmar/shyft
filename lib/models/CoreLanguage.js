
import {
  dataTypes,
  Entity,
} from '../../';


const {
  DataTypeString,
} = dataTypes



export const CoreLanguage = new Entity({
  name: 'language',
  domain: 'core',
  description: 'A language',

  // indexing: {
  //   unique: [
  //     [ 'name' ],
  //     [ 'iso_code' ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      maxLength: 30,
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      maxLength: 30,
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      pattern: '^[a-z]+$',
      minLength: 2,
      maxLength: 2,
      required: true
    },
  }

})

