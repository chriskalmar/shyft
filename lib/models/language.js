
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'language',
      description: 'a language',

      cascadeOnDelete: true,

      attributes: [

        {
          name: 'name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'native_name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'iso_code',
          type: 'string',
          description: 'ISO 639-1 code of the language',
          maxLength: 2,
          required: true
        },

      ],

      indexing: {
        unique: [
          [ 'name' ],
          [ 'iso_code' ],
        ]
      }

    }

  ]

}
