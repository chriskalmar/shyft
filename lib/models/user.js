
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'user',
      description: 'a user',

      isUserTable: true,
      cascadeOnDelete: true,

      attributes: [

        {
          name: 'username',
          type: 'string',
          description: 'Username',
          pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
          minLength: 3,
          maxLength: 30,
          required: true
        },

        {
          name: 'first_name',
          type: 'string',
          description: 'First name',
          maxLength: 30
        },

        {
          name: 'last_name',
          type: 'string',
          description: 'Last name',
          maxLength: 30
        },

        {
          name: 'language',
          type: 'reference',
          description: 'User language',
          target: 'core::language',
          required: true
        },

        {
          name: 'email',
          type: 'email',
          description: 'Email address',
          required: true
        },

        {
          name: 'password',
          type: 'password',
          description: 'User password',
          required: true
        }
      ],

      indexing: {
        unique: [
          [ 'username' ],
          [ 'email' ],
        ]
      }

    }

  ]

}
