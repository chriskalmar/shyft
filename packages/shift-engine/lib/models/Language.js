
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import Index, { INDEX_UNIQUE } from '../index/Index';


export const Language = new Entity({
  name: 'Language',
  description: 'A language',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'name' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'isoCode' ],
    }),
  ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      required: true
    },
  }

})

