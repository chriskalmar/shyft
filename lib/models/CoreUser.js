
import {
  datatypes,
  Entity,
} from '../../';

import CoreLanguage from './CoreLanguage';

const {
  DataTypeString,
} = datatypes

export const CoreUser = new Entity({
  name: 'user',
  domain: 'core',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  // indexing: {
  //   unique: [
  //     [ 'username' ],
  //     [ 'email' ],


  attributes: {

    username: {
      type: DataTypeString,
      description: 'Username',
      pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
      minLength: 3,
      maxLength: 30,
      required: true
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
      maxLength: 30
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
      maxLength: 30
    },

    language: {
      type: CoreLanguage,
      description: 'User language',
      required: true
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true
    }
  }

})



