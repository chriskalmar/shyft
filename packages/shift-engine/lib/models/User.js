
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import Index, { INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';


export const User = new Entity({
  name: 'User',
  domain: 'core',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexing: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],


  attributes: {

    username: {
      type: DataTypeString,
      description: 'Username',
      pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
      minLength: 3,
      maxLength: 30,
      required: true
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
      maxLength: 30
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
      maxLength: 30
    },

    language: {
      type: Language,
      description: 'User language',
      required: true
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true
    }
  }

})



