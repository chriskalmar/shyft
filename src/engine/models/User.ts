import { Entity } from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import { Index, INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';

export const User = new Entity({
  name: 'User',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],

  attributes: {
    username: {
      type: DataTypeString,
      description: 'Username',
      required: true,
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
    },

    language: {
      type: Language,
      description: 'User language',
      required: true,
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true,
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true,
    },
  },
});
