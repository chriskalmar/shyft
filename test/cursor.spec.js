import './setupAndTearDown';
import {
  find,
} from './db';

import {
  asAdmin,
  removeListDynamicData,
} from './testUtils';

import { Profile } from './models/Profile';


const orderByUsernameAsc = {
  orderBy: [{
    attribute: 'username',
    direction: 'ASC'
  }]
}

const orderByUsernameDesc = {
  orderBy: [{
    attribute: 'username',
    direction: 'DESC'
  }]
}

describe('cursor', () => {

  it('after', async () => {

    const cursor = {
      after: {
        Profile: [
          [
            'username',
            'amalia943'
          ],
          [
            'id',
            65
          ]
        ]
      }
    }

    const result = await find(Profile, { ...orderByUsernameAsc, ...cursor }, asAdmin())
    result.data = removeListDynamicData(Profile, result.data)

    expect(result).toMatchSnapshot()
  })


  it('before', async () => {

    const cursor = {
      before: {
        Profile: [
          [
            'username',
            'amalia943'
          ],
          [
            'id',
            65
          ]
        ]
      }
    }

    const result = await find(Profile, { ...orderByUsernameAsc, ...cursor }, asAdmin())
    result.data = removeListDynamicData(Profile, result.data)

    expect(result).toMatchSnapshot()
  })


  it('after + filter', async () => {

    const cursor = {
      after: {
        Profile: [
          [
            'username',
            'ella377'
          ],
          [
            'id',
            103
          ]
        ]
      }
    }

    const filter = {
      username: {
        $starts_with: 'e',
        $not_ends_with: '3',
      }
    }

    const result = await find(Profile, { ...orderByUsernameAsc, ...cursor, filter, first: 4 }, asAdmin())
    result.data = removeListDynamicData(Profile, result.data)
    // console.log(JSON.stringify(result, null, 2));
    expect(result).toMatchSnapshot()
  })


  it('before + filter', async () => {

    const cursor = {
      before: {
        Profile: [
          [
            'username',
            'emilia189'
          ],
          [
            'id',
            90
          ]
        ]
      }
    }

    const filter = {
      username: {
        $starts_with: 'e',
        $not_ends_with: '3',
      }
    }

    const result = await find(Profile, { ...orderByUsernameAsc, ...cursor, filter, first: 4 }, asAdmin())
    result.data = removeListDynamicData(Profile, result.data)
    // console.log(JSON.stringify(result, null, 2));
    expect(result).toMatchSnapshot()
  })


  it('after + before + filter + offset', async () => {

    const cursor = {
      after: {
        Profile: [
          [
            'username',
            'estell684'
          ],
          [
            'id',
            70
          ]
        ]
      },
      before: {
        Profile: [
          [
            'username',
            'edward969'
          ],
          [
            'id',
            8
          ]
        ]
      }
    }

    const filter = {
      username: {
        $starts_with: 'e',
        $not_ends_with: '3',
      }
    }

    let result

    result = await find(Profile, { ...orderByUsernameDesc, ...cursor, filter, offset: 3, first: 2 }, asAdmin())
    result.data = removeListDynamicData(Profile, result.data)
    expect(result).toMatchSnapshot()

    result = await find(Profile, { ...orderByUsernameDesc, ...cursor, filter, offset: 3, last: 2 }, asAdmin())
    result.data = removeListDynamicData(Profile, result.data)
    expect(result).toMatchSnapshot()
  })


})
