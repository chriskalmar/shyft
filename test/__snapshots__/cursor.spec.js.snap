// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cursor after + before + filter + offset 1`] = `
Object {
  "data": Array [
    Object {
      "confirmedAt": null,
      "id": "98",
      "password": "c632a6ac63ec64dd2bec50c1535e96d21c7ab3335eb7a5c488326a3ec6c6e430",
      "username": "elza510",
    },
    Object {
      "confirmedAt": null,
      "id": "103",
      "password": "f056b8cae37dc3ed60ee52da9a2951f8a6c531fe3222a3ad6835dbdd905c7d2a",
      "username": "ella377",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPreviousPage": false,
  },
}
`;

exports[`cursor after + before + filter + offset 2`] = `
Object {
  "data": Array [
    Object {
      "confirmedAt": null,
      "id": "90",
      "password": "469cc643a19fb75901e5d331b8dc14befa8b98810b246c2fb22fff0c382dc39e",
      "username": "emilia189",
    },
    Object {
      "confirmedAt": null,
      "id": "53",
      "password": "96844bc86a7b7c7f058c0a6ae141c162f33ddfca412f2d5899ead8ee713d7418",
      "username": "emelie578",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": false,
    "hasPreviousPage": true,
  },
}
`;

exports[`cursor after + filter 1`] = `
Object {
  "data": Array [
    Object {
      "confirmedAt": null,
      "id": "98",
      "password": "c632a6ac63ec64dd2bec50c1535e96d21c7ab3335eb7a5c488326a3ec6c6e430",
      "username": "elza510",
    },
    Object {
      "confirmedAt": null,
      "id": "53",
      "password": "96844bc86a7b7c7f058c0a6ae141c162f33ddfca412f2d5899ead8ee713d7418",
      "username": "emelie578",
    },
    Object {
      "confirmedAt": null,
      "id": "90",
      "password": "469cc643a19fb75901e5d331b8dc14befa8b98810b246c2fb22fff0c382dc39e",
      "username": "emilia189",
    },
    Object {
      "confirmedAt": null,
      "id": "12",
      "password": "8405ae6de3ebde2abffb2531d437a163a31dc42a109c7369b66b93258c7e1d67",
      "username": "emmet741",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPreviousPage": false,
  },
}
`;

exports[`cursor after 1`] = `
Object {
  "data": Array [
    Object {
      "confirmedAt": null,
      "id": "39",
      "password": "928748d0d4126492733ed8f2c522ed8fa84aa642c1c6a04e7d641476bfbd3e0c",
      "username": "angeline447",
    },
    Object {
      "confirmedAt": null,
      "id": "1",
      "password": "8fed8061e1e33d96169b9026605172a8fe21eb1ce51756187195be0330536fb7",
      "username": "annabel711",
    },
    Object {
      "confirmedAt": null,
      "id": "84",
      "password": "49321648dc8383b8ba73c037dbcdb14873222077d3709bffe454100e5f809773",
      "username": "anne750",
    },
    Object {
      "confirmedAt": null,
      "id": "69",
      "password": "f056b8cae37dc3ed60ee52da9a2951f8a6c531fe3222a3ad6835dbdd905c7d2a",
      "username": "antone347",
    },
    Object {
      "confirmedAt": null,
      "id": "46",
      "password": "9dab09bbccf2cc2c187d7df4c5fc5326ebb050aac7d351284a079529e6e5637b",
      "username": "antonia919",
    },
    Object {
      "confirmedAt": null,
      "id": "99",
      "password": "863a6b632034d24c2d404f944dde121158683787b89cd065fd00d5fa76f68853",
      "username": "antonina448",
    },
    Object {
      "confirmedAt": null,
      "id": "2",
      "password": "0ba4c26f03ae3c7c7a85bf4204539884e52a0e1e511aeaf11d3b2fd627b5957d",
      "username": "anya543",
    },
    Object {
      "confirmedAt": null,
      "id": "26",
      "password": "10a3d43aba9613d935422f749c8cb2751db62090e246894a3e96407498e5ce22",
      "username": "araceli794",
    },
    Object {
      "confirmedAt": null,
      "id": "34",
      "password": "80c980ff9125cd274a3a1cd88cd9dfa4ba0bec9c6336182c78095a90db65b1de",
      "username": "arlie128",
    },
    Object {
      "confirmedAt": null,
      "id": "15",
      "password": "09b47655056ff5f96694fd9ea16b86998c853046ca7adc60587fa152c02f7b4b",
      "username": "athena557",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPreviousPage": false,
  },
}
`;

exports[`cursor before + filter 1`] = `
Object {
  "data": Array [
    Object {
      "confirmedAt": null,
      "id": "8",
      "password": "5f53b0a2be22e28dc4b056f663910c046a33f9b2d779eaf997ea771fe1881a13",
      "username": "edward969",
    },
    Object {
      "confirmedAt": null,
      "id": "58",
      "password": "4d8e7438a0573b2895ab0f49fe82596b564e8ba985bc8be657ac764bbd0d7895",
      "username": "eladio832",
    },
    Object {
      "confirmedAt": null,
      "id": "103",
      "password": "f056b8cae37dc3ed60ee52da9a2951f8a6c531fe3222a3ad6835dbdd905c7d2a",
      "username": "ella377",
    },
    Object {
      "confirmedAt": null,
      "id": "98",
      "password": "c632a6ac63ec64dd2bec50c1535e96d21c7ab3335eb7a5c488326a3ec6c6e430",
      "username": "elza510",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPreviousPage": false,
  },
}
`;

exports[`cursor before 1`] = `
Object {
  "data": Array [
    Object {
      "confirmedAt": null,
      "id": "24",
      "password": "f22e89c251e50eec7c9e184f5584b99fdf48fdbde7317acd6388ce05ea691a43",
      "username": "aida285",
    },
    Object {
      "confirmedAt": null,
      "id": "31",
      "password": "5ad1ebea0aafeb88c3114324ceb91db18605e498e476a787b5c2d2aec09d31e0",
      "username": "alice157",
    },
    Object {
      "confirmedAt": null,
      "id": "42",
      "password": "72ddd2cc4594aa1134b8eaa9da845a31e70ad31ef81b0968f6d58b310c9a9926",
      "username": "alva280",
    },
    Object {
      "confirmedAt": null,
      "id": "72",
      "password": "87143c4a0ff67290eb9f0b0098e015e27dc06bd527a5010ccd04265b69320bcc",
      "username": "alvera707",
    },
  ],
  "pageInfo": Object {
    "hasNextPage": false,
    "hasPreviousPage": false,
  },
}
`;
